src/terminal.rs

```diff
@@ -1,22 +1,46 @@
 #[cfg(unix)]
-pub(crate) fn acquire_terminal(interactive: bool) {
-    use nix::{
-        errno::Errno,
-        sys::signal::{signal, SigHandler, Signal},
-        unistd,
-    };
-    use std::io::IsTerminal;
+use std::{
+    io::IsTerminal,
+    sync::atomic::{AtomicI32, Ordering},
+};
+
+#[cfg(unix)]
+use nix::{
+    errno::Errno,
+    libc,
+    sys::signal::{self, signal, SaFlags, SigAction, SigHandler, SigSet, Signal},
+    unistd::{self, Pid},
+};
+
+#[cfg(unix)]
+static INITIAL_PGID: AtomicI32 = AtomicI32::new(-1);
 
+#[cfg(unix)]
+pub(crate) fn acquire_terminal(interactive: bool) {
     if interactive && std::io::stdin().is_terminal() {
         // see also: https://www.gnu.org/software/libc/manual/html_node/Initializing-the-Shell.html
 
-        take_control();
+        let initial_pgid = take_control();
+
+        INITIAL_PGID.store(initial_pgid.into(), Ordering::Relaxed);
 
         unsafe {
-            // SIGINT and SIGQUIT have special handling above
+            if libc::atexit(restore_terminal) != 0 {
+                eprintln!("ERROR: failed to set exit function");
+                std::process::exit(1);
+            };
+
+            // SIGINT and SIGQUIT have special handling
             signal(Signal::SIGTSTP, SigHandler::SigIgn).expect("signal ignore");
             signal(Signal::SIGTTIN, SigHandler::SigIgn).expect("signal ignore");
             signal(Signal::SIGTTOU, SigHandler::SigIgn).expect("signal ignore");
+            signal_hook::low_level::register(signal_hook::consts::SIGTERM, || {
+                restore_terminal();
+                // TODO: POSIX specifies that if terminated by a signal, exit code should be > 128
+                // Typically, this is 128 + signal_num
```

Comment:
> Should we implement this in this PR?

I'm not sure. I left it as a TODO, because I don't know how we approach POSIX compatibility. We also allow the user to specify arbitrary exit codes via `exit`, and those may overlap with the `128 + signal num` range.

Also, if we stick with a singular exit code, perhaps it should be `exit(1)` instead of `exit(0)` ?

> Follow up question: I saw in `man atexit` that those handlers are not called if terminated by a signal. Would we need to do the same for other terminating signals or can we ignore this if we are for example given a `SIGKILL`?

I guess technically we should, but there are a lot of terminating signals and many are for exceptional or non-applicable cases (e.g., SIGSEGV, SIGALRM, or SIGPROF). The main ones I thought were relevant are `SIGTERM`, `SIGKILL`, and `SIGHUP`. `SIGKILL` can't be caught, and `SIGHUP` indicates that the terminal was closed, so we don't need to restore it. So, I ended up only adding the `SIGTERM` hook, which is what `fish` does as well.

Except that `fish` also catches `SIGPIPE`, which is one I missed. I'm not familiar with how nushell does piping, so this might not be relevant either.

Reply:
> I don't know how we approach POSIX compatibility. We also allow the user to specify arbitrary exit codes via `exit`, and those may overlap with the `128 + signal num` range.

While internally we don't comply with POSIX, I think we can be good citizens to the outside world if they look at us as just a process. So if there is a generally accepted exit code for that we can use it. That our users can commit their own crimes shouldn't stop us :D

> The main ones I thought were relevant are `SIGTERM`, `SIGKILL`, and `SIGHUP`. `SIGKILL` can't be caught, and `SIGHUP` indicates that the terminal was closed, so we don't need to restore it.

Thanks for the explanation, sounds reasonable what you are doing.

> Except that `fish` also catches `SIGPIPE`, which is one I missed. I'm not familiar with how nushell does piping, so this might not be relevant either.

Mhh, yeah not sure how robustly we are handling it when dealing with pipes into/out of the `nu` binary. I think we can deal with this separately for now.

Reply:
Did some more research, and another option would be to re-`raise` the signal as shown in the [manual](https://www.gnu.org/software/libc/manual/html_node/Termination-in-Handler.html). This is what fish does as well for `SIGTERM`.

Reply:
Ah that sounds elegant! So as long as the OS is POSIX compliant we should be good and don't need to do any signal id arithmetic.